File: tests/unit/test_frame_parser.cpp
#include <catch2/catch_test_macros.hpp>
#include "frame_parser.hpp"
#include <vector>

// Builds a valid STREAM_DATA frame:
// [0xAD][0x01][LEN_HI][LEN_LO][ch_id][seq_hi][seq_lo][count_hi][count_lo][samples...]
static std::vector<uint8_t> make_stream_frame(uint8_t ch_id, uint16_t seq,
                                              std::vector<uint16_t> samples)
{
    uint16_t             payload_len = uint16_t(5 + samples.size() * 2);
    std::vector<uint8_t> buf;
    buf.push_back(0xAD);
    buf.push_back(0x01);
    buf.push_back(uint8_t(payload_len >> 8));
    buf.push_back(uint8_t(payload_len));
    buf.push_back(ch_id);
    buf.push_back(uint8_t(seq >> 8));
    buf.push_back(uint8_t(seq));
    buf.push_back(uint8_t(samples.size() >> 8));
    buf.push_back(uint8_t(samples.size()));
    for (uint16_t s : samples) {
        buf.push_back(uint8_t(s >> 8));
        buf.push_back(uint8_t(s));
    }
    return buf;
}

// Builds a frame with an arbitrary type byte and raw payload (for unknown-type tests)
static std::vector<uint8_t> make_frame(uint8_t type, std::vector<uint8_t> payload)
{
    std::vector<uint8_t> buf;
    buf.push_back(0xAD);
    buf.push_back(type);
    buf.push_back(uint8_t(payload.size() >> 8));
    buf.push_back(uint8_t(payload.size()));
    buf.insert(buf.end(), payload.begin(), payload.end());
    return buf;
}

TEST_CASE("parses a single valid STREAM_DATA frame", "[frame_parser]")
{
    FrameParser p;
    auto        frames = p.feed(make_stream_frame(0, 0, {100, 200, 300}));
    REQUIRE(frames.size() == 1);
    REQUIRE(frames[0].channel_id == 0);
    REQUIRE(frames[0].seq == 0);
    REQUIRE(frames[0].samples == std::vector<uint16_t>{100, 200, 300});
}

TEST_CASE("channel_id is captured correctly", "[frame_parser]")
{
    FrameParser p;
    auto        frames = p.feed(make_stream_frame(7, 0, {1}));
    REQUIRE(frames.size() == 1);
    REQUIRE(frames[0].channel_id == 7);
}

TEST_CASE("parses multiple back-to-back frames", "[frame_parser]")
{
    FrameParser p;
    auto        buf = make_stream_frame(0, 0, {1, 2});
    auto        f1  = make_stream_frame(0, 1, {3, 4});
    buf.insert(buf.end(), f1.begin(), f1.end());
    auto frames = p.feed(buf);
    REQUIRE(frames.size() == 2);
    REQUIRE(frames[0].seq == 0);
    REQUIRE(frames[1].seq == 1);
}

TEST_CASE("buffers partial frame across feeds", "[frame_parser]")
{
    FrameParser p;
    auto        bytes   = make_stream_frame(0, 0, {42});
    auto        frames1 = p.feed(std::span(bytes.data(), 4));
    REQUIRE(frames1.empty());
    auto frames2 = p.feed(std::span(bytes.data() + 4, bytes.size() - 4));
    REQUIRE(frames2.size() == 1);
    REQUIRE(frames2[0].samples[0] == 42);
}

TEST_CASE("resyncs after leading garbage bytes", "[frame_parser]")
{
    FrameParser          p;
    std::vector<uint8_t> buf   = {0x01, 0xFF, 0x00};
    auto                 frame = make_stream_frame(0, 0, {7, 8});
    buf.insert(buf.end(), frame.begin(), frame.end());
    auto frames = p.feed(buf);
    REQUIRE(frames.size() == 1);
    REQUIRE(frames[0].samples == std::vector<uint16_t>{7, 8});
}

TEST_CASE("skips unknown frame type using LENGTH field", "[frame_parser]")
{
    FrameParser p;
    // Unknown type 0x00 (DEVICE_INFO) with some payload, followed by a STREAM_DATA
    auto unknown = make_frame(0x00, {0x00, 0x01, 0x0F, 0x01});
    auto stream  = make_stream_frame(0, 5, {99});
    unknown.insert(unknown.end(), stream.begin(), stream.end());
    auto frames = p.feed(unknown);
    REQUIRE(frames.size() == 1);
    REQUIRE(frames[0].seq == 5);
    REQUIRE(frames[0].samples[0] == 99);
}

TEST_CASE("skips multiple unknown types before STREAM_DATA", "[frame_parser]")
{
    FrameParser p;
    auto        buf = make_frame(0x10, {0x80, 0x00});                               // ACK frame
    auto        f2  = make_frame(0x02, {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06}); // PERIODIC_DATA
    auto        f3  = make_stream_frame(2, 0, {55});
    buf.insert(buf.end(), f2.begin(), f2.end());
    buf.insert(buf.end(), f3.begin(), f3.end());
    auto frames = p.feed(buf);
    REQUIRE(frames.size() == 1);
    REQUIRE(frames[0].channel_id == 2);
    REQUIRE(frames[0].samples[0] == 55);
}

TEST_CASE("detects seq drop", "[frame_parser]")
{
    FrameParser p;
    auto        buf = make_stream_frame(0, 0, {1});
    auto        f3  = make_stream_frame(0, 3, {2}); // seq 1 and 2 missing
    buf.insert(buf.end(), f3.begin(), f3.end());
    p.feed(buf);
    REQUIRE(p.seq_drops() == 2);
}

TEST_CASE("no drop for sequential frames", "[frame_parser]")
{
    FrameParser p;
    for (uint16_t i = 0; i < 8; ++i)
        p.feed(make_stream_frame(0, i, {i}));
    REQUIRE(p.seq_drops() == 0);
}

TEST_CASE("seq wraps from 0xFFFF to 0x0000 without false drop", "[frame_parser]")
{
    FrameParser p;
    auto        buf = make_stream_frame(0, 0xFFFE, {1});
    auto        f1  = make_stream_frame(0, 0xFFFF, {2});
    auto        f2  = make_stream_frame(0, 0x0000, {3});
    buf.insert(buf.end(), f1.begin(), f1.end());
    buf.insert(buf.end(), f2.begin(), f2.end());
    p.feed(buf);
    REQUIRE(p.seq_drops() == 0);
}

File: tests/tools/test_frame_parser.py
import struct
from tools.monitor.monitor import (
    FrameParser, StreamDataFrame, AckFrame,
    MAGIC, FrameType,
)


def _frame(type_byte: int, payload: bytes) -> bytes:
    return bytes([MAGIC, type_byte, len(payload) >> 8, len(payload) & 0xFF]) + payload


def make_stream(seq: int, samples: list[int], ch_id: int = 0) -> bytes:
    count = len(samples)
    pl = bytes([ch_id, seq >> 8, seq & 0xFF, count >> 8, count & 0xFF])
    pl += struct.pack(f">{count}H", *samples)
    return _frame(FrameType.STREAM_DATA, pl)


def make_ack(cmd_type: int, status: int) -> bytes:
    return _frame(FrameType.ACK, bytes([cmd_type, status]))


def test_stream_single_frame():
    p = FrameParser()
    frames = p.feed(make_stream(0, [100, 200, 300]))
    assert len(frames) == 1
    f = frames[0]
    assert isinstance(f, StreamDataFrame)
    assert f.channel_id == 0
    assert f.seq == 0
    assert f.samples == [100, 200, 300]


def test_stream_channel_id():
    p = FrameParser()
    frames = p.feed(make_stream(0, [1], ch_id=7))
    assert frames[0].channel_id == 7


def test_stream_multiple_back_to_back():
    p = FrameParser()
    data = make_stream(0, [1, 2]) + make_stream(1, [3, 4])
    frames = p.feed(data)
    assert len(frames) == 2
    assert frames[0].seq == 0
    assert frames[1].seq == 1


def test_stream_partial_buffered():
    p    = FrameParser()
    data = make_stream(0, [10, 20])
    assert p.feed(data[:4]) == []
    frames = p.feed(data[4:])
    assert len(frames) == 1
    assert frames[0].samples == [10, 20]


def test_stream_no_drop_first_frame():
    p = FrameParser()
    p.feed(make_stream(500, [1]))
    assert p.seq_drops == 0


def test_stream_seq_drop_counted():
    p = FrameParser()
    p.feed(make_stream(0, [1]))
    p.feed(make_stream(3, [2]))
    assert p.seq_drops == 2


def test_stream_seq_wrap():
    p = FrameParser()
    p.feed(make_stream(0xFFFE, [1]))
    p.feed(make_stream(0xFFFF, [2]))
    p.feed(make_stream(0x0000, [3]))
    assert p.seq_drops == 0


def test_stream_samples_preserved():
    samples = list(range(256))
    p       = FrameParser()
    frames  = p.feed(make_stream(0, samples))
    assert frames[0].samples == samples


def test_ack_parsed():
    p      = FrameParser()
    frames = p.feed(make_ack(0x81, 0x00))
    assert len(frames) == 1
    f = frames[0]
    assert isinstance(f, AckFrame)
    assert f.cmd_type == 0x81
    assert f.status == 0x00


def test_unknown_type_skipped_via_length():
    p     = FrameParser()
    unk   = _frame(0x05, b"\x01\x02\x03")
    known = make_stream(7, [99])
    frames = p.feed(unk + known)
    assert len(frames) == 1
    assert isinstance(frames[0], StreamDataFrame)
    assert frames[0].seq == 7


def test_multiple_unknown_types_before_known():
    p   = FrameParser()
    buf = _frame(0x20, b"hello") + _frame(0x05, b"") + make_stream(0, [42])
    frames = p.feed(buf)
    assert len(frames) == 1
    assert frames[0].samples == [42]


def test_resyncs_after_garbage():
    p      = FrameParser()
    data   = b"\x00\xff\x12" + make_stream(0, [42])
    frames = p.feed(data)
    assert len(frames) == 1
    assert frames[0].samples == [42]


def test_mixed_types_in_sequence():
    p   = FrameParser()
    buf = make_ack(0x81, 0x00) + make_stream(0, [1, 2, 3])
    frames = p.feed(buf)
    assert len(frames) == 2
    assert isinstance(frames[0], AckFrame)
    assert isinstance(frames[1], StreamDataFrame)
